File: scripts/validate_skills.py
#!/usr/bin/env python3
"""Lightweight Pryzael package checks. Use pinned skills-ref for independent validation."""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SKILLS = ROOT / "skills"
ALLOWED_TOP = {"name", "description", "license", "compatibility", "metadata", "allowed-tools"}
NAME_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
RESOURCE_PATH_RE = re.compile(r"^(?:references|assets|scripts)/[A-Za-z0-9._/-]+$")
INLINE_CODE_RE = re.compile(r"(?<!`)`([^`\n]+)`(?!`)")
MARKDOWN_LINK_RE = re.compile(
    r"\[[^\]\n]*\]\(\s*(?:<([^>\n]+)>|([^\s)]+))(?:\s+['\"][^)\n]*['\"])?\s*\)"
)


def frontmatter(text: str) -> tuple[dict[str, str], list[str]]:
    lines = text.splitlines()
    errors: list[str] = []
    if not lines or lines[0].strip() != "---":
        return {}, ["missing opening YAML frontmatter delimiter"]
    try:
        end = next(i for i in range(1, len(lines)) if lines[i].strip() == "---")
    except StopIteration:
        return {}, ["missing closing YAML frontmatter delimiter"]

    data: dict[str, str] = {}
    current_parent: str | None = None
    for raw in lines[1:end]:
        if not raw.strip() or raw.lstrip().startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip(" "))
        if indent == 0:
            if ":" not in raw:
                errors.append(f"invalid top-level YAML line: {raw!r}")
                continue
            key, value = raw.split(":", 1)
            key = key.strip()
            value = value.strip().strip('"').strip("'")
            data[key] = value
            current_parent = key if not value else None
        elif current_parent == "metadata":
            if ":" not in raw.strip():
                errors.append(f"invalid metadata YAML line: {raw!r}")
        else:
            errors.append(f"unsupported nested YAML under {current_parent!r}: {raw!r}")
    return data, errors


def explicit_resource_references(text: str) -> list[str]:
    """Return only resource paths expressed through established explicit Markdown syntax."""
    references: list[str] = []

    def add(raw: str) -> None:
        candidate = raw.strip()
        candidate = candidate.split("#", 1)[0]
        if RESOURCE_PATH_RE.fullmatch(candidate) and candidate not in references:
            references.append(candidate)

    for match in INLINE_CODE_RE.finditer(text):
        add(match.group(1))

    for match in MARKDOWN_LINK_RE.finditer(text):
        add(match.group(1) or match.group(2))

    return references


def validate_skill(path: Path) -> list[str]:
    errors: list[str] = []
    skill_file = path / "SKILL.md"
    if not skill_file.is_file():
        return ["missing SKILL.md"]

    text = skill_file.read_text(encoding="utf-8")
    data, parse_errors = frontmatter(text)
    errors.extend(parse_errors)

    unknown = set(data) - ALLOWED_TOP
    if unknown:
        errors.append(f"unsupported top-level frontmatter keys: {sorted(unknown)}")

    name = data.get("name", "")
    if name != path.name:
        errors.append(f"name {name!r} does not match directory {path.name!r}")
    if not NAME_RE.fullmatch(name):
        errors.append(f"invalid skill name: {name!r}")
    if len(name) > 64:
        errors.append("name exceeds 64 characters")

    desc = data.get("description", "")
    if not desc:
        errors.append("missing description")
    if len(desc) > 1024:
        errors.append("description exceeds 1024 characters")

    if len(text.splitlines()) > 500:
        errors.append("SKILL.md exceeds 500 lines")

    notice = path / "LICENSE.pstack.txt"
    if not notice.is_file():
        errors.append("missing LICENSE.pstack.txt package notice")

    for resource in explicit_resource_references(text):
        if not (path / resource).exists():
            errors.append(f"broken local resource reference: {resource}")

    return errors


def main() -> int:
    failures = 0
    if not SKILLS.is_dir():
        print("skills directory not found", file=sys.stderr)
        return 2

    for path in sorted(p for p in SKILLS.iterdir() if p.is_dir()):
        errors = validate_skill(path)
        if errors:
            failures += 1
            print(f"FAIL {path.name}")
            for error in errors:
                print(f"  - {error}")
        else:
            print(f"PASS {path.name}")

    return 1 if failures else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate_skills.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { ordinalCompare } from "./deterministic_order.mjs";
import { parseCanonicalSkillMarkdown } from "./skill_package.mjs";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const SKILLS = path.join(ROOT, "skills");
const ALLOWED_TOP = new Set(["name", "description", "license", "compatibility", "metadata", "allowed-tools"]);
const NAME_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const RESOURCE_PATH_RE = /^(?:references|assets|scripts)\/[A-Za-z0-9._/-]+$/;
const INLINE_CODE_RE = /(?<!`)`([^`\n]+)`(?!`)/g;
const MARKDOWN_LINK_RE = /\[[^\]\n]*\]\(\s*(?:<([^>\n]+)>|([^\s)]+))(?:\s+['"][^)\n]*['"])?\s*\)/g;

export function explicitResourceReferences(text) {
  const references = [];

  function add(raw) {
    const candidate = raw.trim().split("#", 1)[0];
    if (RESOURCE_PATH_RE.test(candidate) && !references.includes(candidate)) {
      references.push(candidate);
    }
  }

  for (const match of text.matchAll(INLINE_CODE_RE)) add(match[1]);
  for (const match of text.matchAll(MARKDOWN_LINK_RE)) add(match[1] ?? match[2]);
  return references;
}

export function validateSkill(skillDir) {
  const errors = [];
  const directoryName = path.basename(skillDir);
  const skillFile = path.join(skillDir, "SKILL.md");
  if (!fs.existsSync(skillFile) || !fs.statSync(skillFile).isFile()) {
    return ["missing SKILL.md"];
  }

  const text = fs.readFileSync(skillFile, "utf8");
  const parsed = parseCanonicalSkillMarkdown(text);
  errors.push(...parsed.errors);

  const unknown = Object.keys(parsed.entries)
    .filter((key) => !ALLOWED_TOP.has(key))
    .sort(ordinalCompare);
  if (unknown.length > 0) {
    errors.push(`unsupported top-level frontmatter keys: ${JSON.stringify(unknown)}`);
  }

  const name = parsed.semanticFields.name ?? "";
  if (name !== directoryName) {
    errors.push(`name ${JSON.stringify(name)} does not match directory ${JSON.stringify(directoryName)}`);
  }
  if (!NAME_RE.test(name)) {
    errors.push(`invalid skill name: ${JSON.stringify(name)}`);
  }
  if (name.length > 64) {
    errors.push("name exceeds 64 characters");
  }

  const description = parsed.semanticFields.description ?? "";
  if (!description) {
    errors.push("missing description");
  }
  if (description.length > 1024) {
    errors.push("description exceeds 1024 characters");
  }

  if (text.split(/\r\n|\n|\r/).length > 500) {
    errors.push("SKILL.md exceeds 500 lines");
  }

  const notice = path.join(skillDir, "LICENSE.pstack.txt");
  if (!fs.existsSync(notice) || !fs.statSync(notice).isFile()) {
    errors.push("missing LICENSE.pstack.txt package notice");
  }

  for (const resource of explicitResourceReferences(text)) {
    if (!fs.existsSync(path.join(skillDir, resource))) {
      errors.push(`broken local resource reference: ${resource}`);
    }
  }

  return errors;
}

export function validateAllSkills(skillsDir = SKILLS) {
  if (!fs.existsSync(skillsDir) || !fs.statSync(skillsDir).isDirectory()) {
    return { status: 2, results: [], fatal: "skills directory not found" };
  }

  const directories = fs.readdirSync(skillsDir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .sort((a, b) => ordinalCompare(a.name, b.name));

  const results = directories.map((entry) => ({
    name: entry.name,
    errors: validateSkill(path.join(skillsDir, entry.name)),
  }));
  return {
    status: results.some((result) => result.errors.length > 0) ? 1 : 0,
    results,
    fatal: null,
  };
}

function main() {
  const result = validateAllSkills();
  if (result.fatal) {
    console.error(result.fatal);
    return result.status;
  }

  for (const item of result.results) {
    if (item.errors.length === 0) {
      console.log(`PASS ${item.name}`);
      continue;
    }
    console.log(`FAIL ${item.name}`);
    for (const error of item.errors) console.log(`  - ${error}`);
  }
  return result.status;
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  process.exitCode = main();
}
